File: lib/domain/field_cultivation/policies/field_cultivation_climate_observed_merge_range_policy.rb
# frozen_string_literal: true

module Domain
  module FieldCultivation
    module Policies
      module FieldCultivationClimateObservedMergeRangePolicy
        module_function

        # @param today [Date]
        # @return [Domain::FieldCultivation::Dtos::FieldCultivationClimateObservedMergeRangeDecision]
        def resolve(cultivation_start_date:, cultivation_end_date:, today:)
          mapper = Domain::FieldCultivation::Mappers::FieldCultivationClimateWeatherPayloadMapper
          observed_start = mapper.coerce_optional_date(cultivation_start_date)
          observed_end = mapper.coerce_optional_date(cultivation_end_date)

          return Dtos::FieldCultivationClimateObservedMergeRangeDecision.skip if observed_start.nil? || observed_end.nil?

          actual_end = [ observed_end, today - 1 ].min
          return Dtos::FieldCultivationClimateObservedMergeRangeDecision.skip if observed_start > actual_end

          Dtos::FieldCultivationClimateObservedMergeRangeDecision.range(
            start_date: observed_start,
            end_date: actual_end
          )
        end
      end
    end
  end
end

File: crates/agrr-domain/src/field_cultivation/policies/field_cultivation_climate_observed_merge_range_policy.rs
use time::Date;

use crate::field_cultivation::dtos::FieldCultivationClimateObservedMergeRangeDecision;
use crate::field_cultivation::mappers::field_cultivation_climate_weather_payload_mapper::coerce_optional_date;

pub fn resolve_observed_merge_range(
    display_start_date: Option<&str>,
    display_end_date: Option<&str>,
    cultivation_start_date: Option<Date>,
    cultivation_end_date: Option<Date>,
    today: Date,
) -> FieldCultivationClimateObservedMergeRangeDecision {
    let _ = (display_start_date, display_end_date);
    let observed_start = cultivation_start_date;
    let observed_end = cultivation_end_date;

    let (Some(observed_start), Some(observed_end)) = (observed_start, observed_end) else {
        return FieldCultivationClimateObservedMergeRangeDecision::skip();
    };

    let actual_end = observed_end.min(today - time::Duration::days(1));
    if observed_start > actual_end {
        return FieldCultivationClimateObservedMergeRangeDecision::skip();
    }

    FieldCultivationClimateObservedMergeRangeDecision::range(observed_start, actual_end)
}

#[cfg(test)]
mod policies_field_cultivation_climate_observed_merge_range_policy_test_inline {
    use super::*;
    include!(concat!(env!("CARGO_MANIFEST_DIR"), "/test/field_cultivation/policies_field_cultivation_climate_observed_merge_range_policy_test.rs"));
}
